### DISCUSSION TITLE
Runnable parameters

### DISCUSSION THREAD

**Discussion by marioaag** [avocado-framework/avocado#4567] [Q&A] [answered]
What is the right syntax to pass arguments to an avocado-instrumented test through a Runnable?
```
Runnable('avocado-instrumented', uri, ???)
```

**Comment by beraldoleal** [comment]
Hi @marioaag , you could pass a `kind`, `uri`, `config`, `args` and `kwargs`. args and kwargs are used by specific runners, And you will find the docstring for each runner on `avocado/core/nrunner.py`. For instance, `ExecRunner` uses args as arguments to be given on the command line to the binary of the uri, and `kwargs` as enviroment variables.

If you are interested on `InstrumentedTest` (`avocado/core/runners/avocado_instrumented.py`), args and kwargs (unless you have tags) are not used. Something like this would be enough:

```python
Runnable('avocado-instrumented', '/path/to/your/instrumented_test.py', config=your_config_dict)
```

Sorry if this is not so well documented, but we are not exporting this, publicly yet, and it is used only in our core. So if you have any specific requirement, an use case, and would like to share in more details with us, we can evaluate how incorporate this on our Job API (that is meant to be more public).

**Reply by marioaag** [reply]
No problem, you guys are very responsive and I really appreciate that!

I have been trying using this format:
```
config={'run.test_parameters': [['host_ip', '300']]}
```
At runnable, suit and job level and looks like it is not getting available using:
```
self.params.get('host_ip')
```
At the test level

What I'm trying to achieve is create an execution sequence base in information coming from yaml file, in the yaml file is specified subsets of tests to be run in parallel or sequential way, and if the same test should be ran multiple times whit different parameter values.

Right now the sequence is getting generated configuring suits and runnables by my self, and is running as expected, the only piece missing is get the parameter specified at runnable level being used at the execution time. This is to be able to run the same test multiple times in parallel using different values for the same parameter.

**Reply by marioaag** [reply]
The parameter is visible in the job.log:
```
2021-05-07 16:45:10,993 job              L0253 I    |  'run.dict_variants': [],
2021-05-07 16:45:10,993 job              L0253 I    |  'run.execution_order': 'variants-per-test',
2021-05-07 16:45:10,993 job              L0253 I    |  'run.test_parameters': [['host_ip', '300']],   <-------- expected parameter
2021-05-07 16:45:10,993 job              L0253 I    |  'run.test_runner': 'nrunner',
```

But trying to access it in the test is returning None:
```
from time import sleep
from avocado import Test

class CheckHostStatus(Test):
    def test_os_info(self):
        """Retrieve OS information via ssh connection"""
        # Retrieving required parameters
        host_ip = self.params.get('host_ip')
        sleep(5)
        self.log.info(f'Host IP: {host_ip}')
        self.log.info("OS Info")
```

Debug logs:
```
16:45:12 INFO | INIT 1-CheckHostStatus.test_os_info
16:45:12 DEBUG| PARAMS (key=timeout, path=*, default=None) => None
16:45:12 DEBUG| Test metadata:
16:45:12 DEBUG|   filename: /Users/marioalvarado/git/iLabConfig/test/avocado/testcases/CheckHostStatus/CheckHostStatus.py
16:45:12 DEBUG|   teststmpdir: /var/folders/pn/xdk10w_12dx8_bvjjrfh35hh0000gn/T/avocado_4ekrn8q2
16:45:12 DEBUG|   workdir: /var/folders/pn/xdk10w_12dx8_bvjjrfh35hh0000gn/T/tmpcsd_ya1y/test-results/tmp_dirjy24070q/1-CheckHostStatus.test_os_info
16:45:12 INFO | START 1-CheckHostStatus.test_os_info
16:45:12 DEBUG| DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
16:45:12 DEBUG| PARAMS (key=host_ip, path=*, default=None) => None
16:45:17 INFO | Host IP: None
16:45:17 INFO | OS Info
16:45:17 INFO | PASS 1-CheckHostStatus.test_os_info
16:45:17 INFO | 
```

**Reply by beraldoleal** [reply]
Are you using nrunner or the default runner?

I did a test with the default runner against our `examples/tests/sleeptest.py` and it worked as expected:

```python3
avocado run tests/sleeptest.py -p sleep_length=10
```

However, using the next runner (`--test-runner=nrunner`), it seems that we have a problem, indeed. I just created #4569 so we can track this.

**Reply by richtja** [reply]
Hi @marioaag If you are trying to run multiple variants of one test, you can use [CIT](https://avocado-framework.readthedocs.io/en/latest/plugins/optional/varianter_cit.html) or [Yaml_to_mux](https://avocado-framework.readthedocs.io/en/latest/plugins/optional/varianter_yaml_to_mux.html) plugins? It might be useful for you.

**Reply by marioaag** [reply]
@beraldoleal, yes I'm using nrunner, one of my principal goals is have a solution that let me run multiple test in parallel.

**Reply by marioaag** [reply]
@richtja, the Yaml_to_mux as I know is not available for nrunner or did something change?

**Reply by richtja** [reply, marked as answer]
Yes you are right, if you are using the nrunner, the varianter plugins are not available right now.
